### DISCUSSION TITLE
Why does List patterns require System.Index

### DISCUSSION THREAD

**Discussion by CreateAndInject** [dotnet/csharplang#5792] [Q&A] [answered]
.NET Framework can't compile this code since CS0518:
```
        static void Main(string[] args)
        {
            if (args is ["a", "b"])
                Console.WriteLine("OK");
        }
```

.NET5 produce this code:
```
private static void Main(string[] args)
{
	if (args != null && args.Length == 2 && args[0] == "a" && args[1] == "b")
	{
		Console.WriteLine("OK");
	}
}
```
There's no dependency to `System.Index`, so why does the compiler require `System.Index`?

**Comment by HaloFour** [comment, marked as answer]
Presumably some forms of list patterns do rely on `System.Index` and `System.Range` and it's better to have the feature be supported in totality rather than some forms of the syntax work and other forms break.  Also, now that C# ships in lock-step with .NET targeting new features on older runtimes isn't supported.

**Reply by 333fred** [reply]
This is correct. To keep the user experience consistent, Index is always required.

**Reply by CreateAndInject** [reply]
> This is correct. To keep the user experience consistent, Index is always required.

It's not friendly to .NET Framework developer, we're always woring on Windows and do not need .NET Core

**Reply by HaloFour** [reply]
Unfortunately that might be true of any new features, particularly those that will rely on runtime changes that aren't going to be backported to .NET Framework.  There are [NuGet packages](https://www.nuget.org/packages/IndexRange/) which will provide compatible implementations of `System.Index` and `System.Range` for polyfill, but the language team has been pretty upfront that this is unsupported territory.

**Reply by CreateAndInject** [reply]
> Unfortunately that might be true of any new features, particularly those that will rely on runtime changes that aren't going to be backported to .NET Framework. There are [NuGet packages](https://www.nuget.org/packages/IndexRange/) which will provide compatible implementations of `System.Index` and `System.Range` for polyfill, but the language team has been pretty upfront that this is unsupported territory.

C# compiler can become more smarter if the compiler team want, since the produced IL doesn't dependent both System.Index & System.Rangle at all.

**Reply by spydacarnage** [reply]
Also, .NET Core / NET5+ is not just about cross-platform compatibility - they also contain a large number of serious performance improvements.  I only ever work on Windows, too, but I still felt the need to upgrade for that alone.

**Reply by HaloFour** [reply]
Yes, the compiler could do that, but at the risk of being a bad user experience and specifically for an unsupported target runtime.

**Reply by CreateAndInject** [reply]
> Presumably some forms of list patterns do rely on `System.Index` and `System.Range`

It's not conflict, compiler know when rely on and when not, so just report error when rely on. I don't think the compiler doesn't know if a code rely on or not when compiling.

**Reply by HaloFour** [reply]
> It's not conflict, compiler know when rely on and when not, so just report error when rely on. I don't think the compiler doesn't know if a code rely on or not when compiling.

Yes, it does, but to have a feature sometimes work and sometimes not is a poor experience.  C# 8.0+ is not supported on .NET Framework and the team is not going to go out of their way to make the experience smoother for unsupported runtimes.

**Reply by 333fred** [reply]
> It's not friendly to .NET Framework developer

As HaloFour indicated, .NET Framework is maintenance mode. While it will be supported for as long as Windows ships it, it is not receiving new features, and that includes new C# language versions. Compiler support for new language features on earlier frameworks is made on a best-effort basis, and we do not support the scenario. If it is easier for the implementation to simply check for `Index` being present in all cases, and it only impacts an unsupported scenario, we will do so. There are several easy ways to get the `Index` types on previous versions of the framework, such as https://www.nuget.org/packages/IndexRange/1.0.1-beta.1.
